### PR TITLE
Better sourcemap handling

### PR DESCRIPTION
**Pull request by kherock**
Hi!

I realize this isn't a very active project, but it would be awesome if this could get merged and published. I really like that this works as a decent alternative to using `css-loader`, as that adds too much overhead and I'd rather configure everything `postcss` myself.

Anyway, there was a lot of extra fluff around the sourcemap handling that made it hard for paths to show up correctly in with both `style-loader` and `ExtractTextPlugin` since the `sourceRoot` was hardcoded to `webpack://`. If users want the CSS source to map under the `webpack://` root, they can use
```javascript
loaders: ['style-loader', 'css-raw-loader?sourceRoot=webpack:///']
```
 I've also made a PR to `style-loader` so that this works too
```javascript
loaders: ['style-loader?sourceRoot=webpack:///', 'css-raw-loader']
```

### CODE REVIEW

**Comment by catdad** [comment]
I basically just ripped off the sourcemap handling from `css-loader`. I'll check out what you did.

File: index.js [hunk 1]
```diff
@@ -1,68 +1,47 @@
...
-                }
-                return "/" + p;
-            }, this);
-            map.sourceRoot = "webpack://";
-        }
-        map = JSON.stringify(map);
+    if (map && config.sourceMap !== false) {
+        return 'module.exports = [[module.id, ' + JSON.stringify(content) + ', "", ' + JSON.stringify(map) + ']];';
+    } else {
+        return 'module.exports = ' + JSON.stringify(content) + ';';
```
Comment (catdad) [inline index.js:44]: I'm confused. I remember needing to wrap the CSS content in all that JavaScript scruft in order for it to load. Did that change? What does this mean for the change? Sorry, I have not been following Webpack development.

File: index.js [hunk 2]
```diff
@@ -1,68 +1,47 @@
...
-        var sourceMap = new SourceListMap();
-        sourceMap.add(content, cssRequest, content);
-        map = sourceMap.toStringWithSourceMap({
-            file: request
-        }).map;
+    if (map && map.sources) {
+        var context = config.context || this.options.context;
+        var requestShortener = new RequestShortener(context);
+        map.sources = map.sources.map(function (source) {
+            return requestShortener.shorten(path.resolve(context, source));
```
Comment (catdad) [inline index.js]: I am also not sure what this does. Can you elaborate, please?

File: index.js [hunk 3]
```diff
@@ -1,68 +1,47 @@
...
-		map = minimizeResult.sourceMap;
-		content = minimizeResult.styles;
-		if(typeof map !== "string") {
-			map = JSON.stringify(map);
+    this.value = content;
+
+    if (config.minimize || this.minimize && config.minimize !== false) {
+        var options = Object.create(config);
+        if (config.sourceMap || this.sourceMap && config.sourceMap !== false) {
+            options.sourceMap = typeof map === 'string' ? map : (map ? JSON.stringify(map) : true);
```
Comment (catdad) [inline index.js:22]: Umm.... what? So `options.sourceMap` can be a random string, a stringified random truthy value, or a boolean? That doesn't seem clean. What is the intent here?

File: index.js (hunk 1 above)
Comment (kherock) [reply index.js:44]: It seems that most loaders (at least `ExtractTextPlugin` and `style-loader`) will accept either a plain string or a list of three-long entries ([see how `style-loader` handles the result here](https://github.com/webpack/style-loader/blob/master/index.js#L25)). I'm pretty sure everything in `css-base.js` was specific to things css-loader was doing. So what this states is if there's a map and sourcemaps aren't explicitly disabled (e.g. `css-raw-loader?-sourceMap`), return the array form with the map. Otherwise give a raw string (à la `raw-loader`).

File: index.js (hunk 2 above)
Comment (kherock) [reply index.js]: This is accomplishing the same thing as before, just a bit more succinctly since it uses webpack's `RequestShortener`. I'm providing webpack with a consistent absolute path to a resource so we get a tidy relative path (e.g. `node_modules/angular-material/angular-material.scss` becomes `~/angular-material/angular-material.scss`).

File: index.js (hunk 3 above)
Comment (kherock) [reply index.js:22]: I left the `sourceRoot` property to be blank by default since it upsets how `ExtractTextPlugin` process maps. If it's set to `webpack://`, `ExtractTextPlugin` still runs the full path through webpack's own `devtoolModuleFilenameTemplate` and the full path ends up being something like `webpack:///./webpack://~/some-module/index.css`.

I suppose we could to a `typeof options.sourceMap === 'string'` check just to enforce a string, but it should be a self-explanatory option. Other loaders such as `babel-loader` and the `UglifyJsPlugin` allow setting `sourceRoot` by a query param or in options.

**Comment by catdad** [comment]
Hey, sorry I disappeared for so long. I basically just forgot about this entirely and didn't remember until I saw the email from that last commit.

So... I thought about this a little bit. I hacked this module together in about 20 minutes when the folks at webpack did not recognize this as a real need (and rejected my PR for the proper css loader). I didn't expect this to actually become a module with an unsettlingly large download amount. I actually no longer use webpack, so I am not sure how to judge the quality of this PR. It seems like you are changing some important stuff, but I have no looked into webpack code in a quite a while (I hope they've started to clean it up... there was some scary stuff in there), so I am not sure if this is all best practice.

So in short, I need to add some tests to this module and make sure that this PR is not breaking any support or causing adverse effects. I will try to look at it this week, but I am not sure if I'll be able to. In the meantime, I guess I could publish this to NPM as an alpha if you need to use it for your builds. If you have the inexplicable desire to help with any of that, let me know.

**Comment by kherock** [comment]
Hey!

It's ok, I would have pressed harder if it was super urgent. I only came across this module originally because it's pretty aptly named ; ).  I've been using my own fork in the meantime, but it would be nice to have my changes available on npm.

I can tell you definitely ripped a decent part of it straight from css-loader—it is basically raw-loader + sourcemaps + cleancss, and I wasn't too happy with how sourcemaps were being handled, so I got rid of a lot of fluff and changed sourcemap handling to my liking.

**Comment by catdad** [comment]
Hey. Again, super late. You may have probably moved on by now. I don't do much front-end development lately, and other projects have been taking up a lot of my time.

Anyway, I have merged this to the [next](https://github.com/catdad/css-raw-loader/tree/next) branch and published it as `1.0.0-alpha.1` for now. You can install this using one of the following:

```bash
npm install css-raw-loader@1.0.0-alpha.1
```

```bash
npm install css-raw-loader@next
```

Currently, it will not install by default. I have not done much to validate any of the functionality, so I do not want to push it out as the default version.

Ideally, this module should have tests, making sure that it functions appropriately. I have no plans to work on this module though. If you are interested in contributing, let me know.
